packages/commerce-sdk-react/src/auth/index.ts

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (c) 2022, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import {helpers, ShopperLogin, ShopperLoginTypes} from 'commerce-sdk-isomorphic'
+import jwtDecode from 'jwt-decode'
+import {ApiClientConfigParams} from '../hooks/types'
+import {BaseStorage, LocalStorage, CookieStorage} from './storage'
+
+type Helpers = typeof helpers
+interface AuthConfig {
+    shortCode: string
+    organizationId: string
+    clientId: string
+    siteId: string
+    redirectURI: string
+    proxy: string
+}
+
+interface JWTHeaders {
+    exp: number
+    iat: number
+}
+
+// this type is slightly different from ShopperLoginTypes.TokenResponse, reasons:
+// 1. TokenResponse is too generic (with & {[key:string]: any}), we need a more
+//    restrictive type to make sure type safe
+// 2. The refresh tokens are stored separately for guest and registered user. Instead
+//    of refresh_token, we have refresh_token_guest and refresh_token_registered
+type AuthDataKeys =
+    | 'access_token'
+    | 'customer_id'
+    | 'enc_user_id'
+    | 'expires_in'
+    | 'id_token'
+    | 'idp_access_token'
+    | 'refresh_token_guest'
+    | 'refresh_token_registered'
+    | 'token_type'
+    | 'usid'
+type AuthDataMap = Record<
+    AuthDataKeys,
+    {
+        storage: BaseStorage
+        key: string
+        callback?: () => void
+    }
+>
+
+/**
+ * This singleton class is used to handle shopper authentication.
```

Comment:
Minor: Is this really a singleton? We can call `new Auth()` as many as we want, and it will result in new instance each time.

Reply:
Technically speaking it's not a singleton... I'll remove the misleading comment...

Reply:
But we do want to use only 1 instance to keep track of the auth pending state for all hooks.